src/neural/onnx/network_onnx.cc

```diff
@@ -98,74 +111,191 @@ class OnnxNetwork : public Network {
   int value_head_ = -1;
   int mlh_head_ = -1;
   NetworkCapabilities capabilities_;
+  bool fp16_;
+  // The batch size to use, or -1 for variable.
+  int batch_size_;
+  static constexpr int max_batch_size_ = 1024;
+  // For conditional locking if running the DML provider.
+  OnnxProvider provider_;
+  std::mutex lock_;
 };
 
-float OnnxComputation::GetQVal(int sample) const {
+template <typename DataType>
+OnnxComputation<DataType>::OnnxComputation(OnnxNetwork* network)
+    : network_(network) {
+  output_tensors_data_.resize(network_->outputs_.size());
+  output_tensors_step_.resize(network_->outputs_.size());
+  output_tensors_step_[network_->policy_head_] = 1858;
+  output_tensors_data_[network_->policy_head_] =
+      std::vector<DataType>(1858 * network_->max_batch_size_);
+  if (network_->wdl_head_ != -1) {
+    output_tensors_step_[network_->wdl_head_] = 3;
+    output_tensors_data_[network_->wdl_head_] =
+        std::vector<DataType>(3 * network_->max_batch_size_);
+  }
+  if (network_->value_head_ != -1) {
+    output_tensors_step_[network_->value_head_] = 1;
+    output_tensors_data_[network_->value_head_] =
+        std::vector<DataType>(network_->max_batch_size_);
+  }
+  if (network_->mlh_head_ != -1) {
+    output_tensors_step_[network_->mlh_head_] = 1;
+    output_tensors_data_[network_->mlh_head_] =
+        std::vector<DataType>(network_->max_batch_size_);
+  }
+}
+
+template <typename DataType>
+void OnnxComputation<DataType>::AddInput(InputPlanes&& input) {
+  raw_input_.emplace_back(input);
+  if (raw_input_.size() > network_->max_batch_size_) {
+    throw Exception("NN input exceeds max batch size of " +
+                    std::to_string(network_->max_batch_size_) + ".");
+  }
+}
+
+template <typename DataType>
+float OnnxComputation<DataType>::GetQVal(int sample) const {
   if (network_->wdl_head_ != -1) {
-    const auto& data =
-        output_tensors_[network_->wdl_head_].GetTensorData<float>();
-    return data[sample * 3 + 0] - data[sample * 3 + 2];
+    const auto& data = output_tensors_data_[network_->wdl_head_];
+    if (std::is_same<Ort::Float16_t, DataType>::value) {
+      return FP16toFP32(data[sample * 3 + 0]) -
+             FP16toFP32(data[sample * 3 + 2]);
+    } else {
+      return data[sample * 3 + 0] - data[sample * 3 + 2];
+    }
   } else {
-    const auto& data =
-        output_tensors_[network_->value_head_].GetTensorData<float>();
-    return data[sample];
+    const auto& data = output_tensors_data_[network_->value_head_];
+    if (std::is_same<Ort::Float16_t, DataType>::value) {
```

Comment:
This construction is repeated several times in the file, consider having a separate function for it.
Maybe even not a template one, but just override. Something like:

```cpp
float AsFloat(float x) { return x; }
float AsFloat(Ort::Float16_t x} { return FP16toFP32(x); }
```